### PR TITLE
Protest albums per date

### LINKED ISSUE #185: Create protest albums per date
**How it works now:**
Currently, in the protest page, we show the 6 last pictures uploaded to it. 
When a visitor clicks the `לצפייה בגלריית ההפגנה` button, they see a list with _all_ the pictures uploaded the protest.

**What needs to be changed:**
Instead of showing all the pictures, we should show the picture galleries, sorted by dates. 
Here's a rough design for the galleries:

Screen Shot 2020-11-05 at 12 32 22
On mobile you can change it to be 2 in a row, but feel free to play with it and change the design as you like it.

**How to approach:**
The `PictureGallery` component currently fetches & shows all the picture. 
It needs to be changed so it'll fetch one picture, the first one taken, per date. (I guess the component name needs to be changed too).

Pictures are being saved in the database inside the `pictures` collection. Each picture holds a `protestId` field and a `date` field. 

After you'll fetch a picture for each date, you need to create a new component & route for the gallery itself. 
Inside it, you should fetch all the pictures for the specified `date` & `protestId`, and show them like the pictures are being shown at the moment.

### PR DESCRIPTION
**Pull request by MSanbira**

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -69,9 +69,9 @@ export function getDateString(date) {
   return [year, month, day].join('-');
 }
 
-export function formatDate(dateStr) {
+export function formatDate(dateStr, withYear = false) {
```
Comment (guytepper) [inline src/utils.js:78]: Can you add a JSDoc here, with the parameters types & description, and the expected return value (with an example)

File: src/utils.js [hunk 2]
```diff
@@ -81,6 +81,10 @@ export function dateToDayOfWeek(dateStr) {
   return daysOfWeek[date.getDay()];
 }
 
+export function dateToDayOfWeekAndDate(dateStr) {
```
Comment (guytepper) [inline src/utils.js:95]: Same as above (JSDoc)

File: src/views/ProtestPage.js [hunk 3]
```diff
@@ -160,7 +161,7 @@ function ProtestPageContent({ protest, user, userCoordinates }) {
           </Details>
         </Info>
 
-        {galleryMatch?.isExact ? (
+        {galleryMatch?.isExact || galleryDateMatch?.isExact ? (
           <PictureGallery protestId={protest.id} date={'2020-10-31'} />
```
Comment (guytepper) [inline src/views/ProtestPage.js]: Remove unused date prop
```suggestion
          <PictureGallery protestId={protest.id} />
```

File: src/components/PictureGallery/PictureGallery.js [hunk 4]
```diff
@@ -1,29 +1,59 @@
...
 
-function PictureGallery({ protestId, date }) {
-  const [pictures, setPictures] = useState([]);
+function PictureGallery({ protestId }) {
   const history = useHistory();
   const store = useStore();
+  const route = matchPath(history.location.pathname, {
+    path: '/protest/:id/gallery/:date',
+  });
+  const galleryDate = route ? route.params.date : null;
```
Comment (guytepper) [inline src/components/PictureGallery/PictureGallery.js]: Will this work too?

```suggestion
  const galleryDate = route?.params.date;
```

File: src/components/PictureGallery/PictureGallery.js [hunk 5]
```diff
@@ -41,6 +71,32 @@ function PictureGallery({ protestId, date }) {
...
+function ImageDateSection({ src, date, onClick }) {
+  return (
+    <ImageDateContainer onClick={onClick}>
+      <ImageThumbnail src={src} />
+      <DateTitle>{dateToDayOfWeekAndDate(date)}</DateTitle>
+    </ImageDateContainer>
+  );
+}
+
+function getPictureForEveryDate(pictures) {
```
Comment (guytepper) [inline src/components/PictureGallery/PictureGallery.js]: Can you move these helper functions to `PictureGalleryUtils.js`, under the `PictureGallery` component folder, and add a JSDoc for each of them?

Comment (guytepper) [inline src/components/PictureGallery/PictureGallery.js]: I think `getGalleryThumbnails` would be a clearer name for this method

File: src/components/PictureGallery/PictureGallery.js [hunk 6]
```diff
@@ -41,6 +71,32 @@ function PictureGallery({ protestId, date }) {
...
+  const datesValidation = [];
```
Comment (guytepper) [inline src/components/PictureGallery/PictureGallery.js]: I think `uniqueDates` would be clearer for this

File: src/components/PictureGallery/PictureGallery.js [hunk 7]
```diff
@@ -41,6 +71,32 @@ function PictureGallery({ protestId, date }) {
...
+    if (!datesValidation.includes(picture.eventDate)) {
+      datesValidation.push(picture.eventDate);
+      picturesForEveryDate.push(picture);
+    }
+  });
+
+  return picturesForEveryDate;
+}
+
+function getAllDatePics(pictures, date) {
```
Comment (guytepper) [inline src/components/PictureGallery/PictureGallery.js]: `getPicturesForDate` would be clearer for this 🧼

File: src/components/PictureGallery/PictureGallery.js [hunk 8]
```diff
@@ -75,6 +131,22 @@ const SectionTitle = styled.div`
...
+  display: flex;
+  flex-flow: column nowrap;
+  align-items: center;
+`;
+
+const DateTitle = styled.div`
+  font-size: 16px;
+  line-height: 19px;
+  color: black;
+  margin-top: 10px;
```
Comment (guytepper) [inline src/components/PictureGallery/PictureGallery.js]: Can you move this to the top? (line 142)
We try to follow [idiomatic css](https://github.com/necolas/idiomatic-css) guidelines

File: src/components/PictureGallery/PictureGallery.js [hunk 9]
```diff
@@ -1,29 +1,59 @@
...
-  const [pictures, setPictures] = useState([]);
+function PictureGallery({ protestId }) {
   const history = useHistory();
   const store = useStore();
+  const route = matchPath(history.location.pathname, {
+    path: '/protest/:id/gallery/:date',
+  });
+  const galleryDate = route ? route.params.date : null;
+  const [pictures, setPictures] = useState([]);
+  const [isGotPictures, setIsGotPictures] = useState(false);
```
Comment (guytepper) [inline src/components/PictureGallery/PictureGallery.js]: `picturesFetchState` will be nicer for this 🤓

**Review by guytepper — CHANGES_REQUESTED** [review]
It works great! Thank you.
Just some small adjustments required before we'll merge this :)

File: src/components/PictureGallery/PictureGalleryUtils.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Get a Picture array - first one of each date.
+ * @param {Picture[]} pictures - An array of picture objects.
+ * @returns {Picture[]} - one for of each date
```
Comment (guytepper) [inline src/components/PictureGallery/PictureGalleryUtils.js]: What's `Picture` means here? I know JSDoc understands `array` as a type, how will it interperate `Picture`?

Comment (MSanbira) [reply src/components/PictureGallery/PictureGalleryUtils.js]: To tell the truth, I don't really use JSDoc, and in my work, I use typescript so I thought I'll use a typescript like type. but tell me what's more clear to write.

Comment (guytepper) [reply src/components/PictureGallery/PictureGalleryUtils.js]: We'll start using TypeScript soon, but for now let's stick with the JSDoc spec, which requires to enter just the type.
```suggestion
 * @returns {array} - Array of pictures, one for of each date
```

File: src/components/PictureGallery/PictureGalleryUtils.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Get a Picture array - first one of each date.
+ * @param {Picture[]} pictures - An array of picture objects.
```
Comment (guytepper) [inline src/components/PictureGallery/PictureGalleryUtils.js]:
```suggestion
 * @param {array} pictures - An array of picture objects.
```

**Review by guytepper — APPROVED** [review]

**Pull request merged**